ilastik/applets/batchProcessing/hbp.py

```diff
@@ -0,0 +1,137 @@
+    token_headers = {"Authorization": f"Token {payload.token}"}
+
+    upload_resp = requests.post(
+        cfg["hbp"]["upload_file_url"],
+        data=payload.project,
+        headers={**token_headers, "Content-Disposition": f'attachment; filename="{payload.filename}"'},
+        timeout=timeout,
+    )
+    upload_resp.raise_for_status()
+
+    project_resp = requests.post(
+        cfg["hbp"]["create_project_url"],
+        json={
+            "file": upload_resp.json()["url"],
+            "workflow": payload.workflow,
+            "num_channels": payload.num_channels,
+            "min_block_size_z": 0 if payload.compute_in_2d else payload.min_block_size,
+            "min_block_size_y": payload.min_block_size,
+            "min_block_size_x": payload.min_block_size,
+        },
+        headers=token_headers,
+        timeout=timeout,
+    )
+    project_resp.raise_for_status()
+
+    return project_resp.json()["html_url"]
+
+
+def partial_copy(
+    name: str, obj: Union[h5py.Group, h5py.Dataset], *, dest: h5py.Group, skip_prefix: Optional[str] = None
+) -> None:
+    """Deep copy obj to dest, but skip contents of datasets whose names start with the prefix.
+
+    If `skip_prefix` is ``None``, include all datasets' contents.
+    If `skip_prefix` is an empty string, skip all datasets' contents.
+
+    Intended to be used as a curried callback for :meth:`h5py.Group.visititems`.
+    """
+    if isinstance(obj, h5py.Group):
+        group = dest.create_group(name)
+        group.attrs.update(obj.attrs)
+    elif isinstance(obj, h5py.Dataset) and skip_prefix is not None and name.startswith(skip_prefix):
+        dataset = dest.create_dataset_like(name, obj)
+        dataset.attrs.update(obj.attrs)
+    else:
+        dest.copy(obj, name)
+
+
+def serialize_project_file(project_file: h5py.File, *, include_local_data: bool = False) -> bytes:
+    """Convert project file into byte representation."""
+    skip_prefix = None if include_local_data else "Input Data/local_data"
+    with io.BytesIO() as buf:
+        with h5py.File(buf) as dest:
+            project_file.visititems(functools.partial(partial_copy, dest=dest, skip_prefix=skip_prefix))
+        return buf.getvalue()
+
+
+def num_channels(op: OpDataSelectionGroup) -> int:
+    role_index = op.DatasetRoles.value.index("Raw Data")
+    dataset_info = op.DatasetGroup[0][role_index].value
+    try:
+        return dataset_info.laneShape[dataset_info.axistags.channelIndex]
+    except IndexError:
+        return 1
+
+
+def min_block_size(op: OpFeatureSelection) -> int:
+    selected = numpy.argwhere(op.SelectionMatrix.value)
+    if not selected.size:
+        return 0
+    return max(halo(scale=op.Scales.value[col], order=ORDERS[op.FeatureIds.value[row]]) for row, col in selected)
+
+
+def compute_in_2d(op: OpFeatureSelection) -> bool:
+    selected = numpy.argwhere(op.SelectionMatrix.value)
+    return any(op.ComputeIn2d.value[col] for _row, col in selected)
```

Comment:
I think if we have even one 3D feature, `min_block_size` should be positive because, in order to compute that feature, cloud_ilasik has to do 3D blocking.

Reply:
but you're checking 2D-ness. So all features are 2D is the same as none is 3D. I still don't see how any is playing in here.

But yes, one could argue if blocking of z=0 and z=1 should be a different thing at all.

Reply:
OK, so I've looked up how we are using this data in cloud_ilasik. Currently, it is used only for [filtering datasets for a particular project](https://github.com/ilastik/cloud_ilastik/blob/900e66fb2621b897f38f574d2f132c976dc1e716/batch/views.py#L71).

Therefore, if we have a classifier that has been trained with even a single 3D feature, using that classifier requires the dataset to be at least `min_block_size_z` pixels along the Z axis.